akka-remote/src/main/scala/akka/remote/artery/RemoteInstrument.scala

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.remote.artery
+
+import akka.actor.{ ActorRef, ExtendedActorSystem }
+import akka.util.{ ByteString, OptionVal }
+
+/**
+ * INTERNAL API
+ *
+ * Part of the monitoring SPI which allows attaching metadata to outbound remote messages,
+ * and reading in metadata from incoming messages.
+ *
+ * Multiple instruments are automatically handled, however they MUST NOT overlap in their idenfitiers.
+ */
+abstract class RemoteInstrument {
+  /**
+   * Instrument identifier.
+   *
+   * MUST be >=0 and <32.
+   *
+   * Values between 0 and 7 are reserved for Akka internal use.
+   */
+  def identifier: Byte
+
+  /**
+   * Called right before putting the message onto the wire.
+   * Parameters MAY be `null` (except `message`)!
+   *
+   * @return `metadata` rendered to be serialized into the remove envelope, or `null` if no metadata should be attached
+   */
+  def remoteMessageSent(recipient: ActorRef, message: Object, sender: ActorRef): ByteString
+
+  /**
+   * Called once a message (containing a metadata field designated for this instrument) has been deserialized from the wire.
+   */
+  def remoteMessageReceived(recipient: ActorRef, message: Object, sender: ActorRef, metadata: ByteString): Unit
```

Comment:
If we have the `ByteBuffer` API, then we might not need the specialized `MetadataMap`?